[GStreamer] Fix media player's duration calculation

eocanha opened the pull request:
#### 0d46fa12c99c41a159a9ae5d394a032ec09fd165
<pre>
[GStreamer] Fix media player&apos;s duration calculation
<a href="https://bugs.webkit.org/show_bug.cgi?id=245148">https://bugs.webkit.org/show_bug.cgi?id=245148</a>

Reviewed by Xabier Rodriguez-Calvar.

Videos without an initially known duration, such as <a href="https://traffic.megaphone.fm/GLT1153394917.mp3?updated=1617734730">https://traffic.megaphone.fm/GLT1153394917.mp3?updated=1617734730</a>,
are currently detected as infinite videos (live streams). They should be detected as
fixed-length videos instead of as a live broadcast.

This patch changes the behaviour to return invalid time when the video is known not
to be live.

Based on a patch by: Pawel Lampe &lt;pawel.lampe@gmail.com&gt;
See: <a href="https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/923">https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/923</a>

* Source/WTF/wtf/PrintStream.h:
(WTF::boolForPrinting): Added overloaded function to print std::optional&lt;bool&gt; values.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::seek): Migrate to optional m_isLiveStream preserving the previous behaviour (unset means false).
(WebCore::MediaPlayerPrivateGStreamer::setRate): ditto.
(WebCore::MediaPlayerPrivateGStreamer::setPreload): ditto.
(WebCore::MediaPlayerPrivateGStreamer::buffered const): ditto.
(WebCore::MediaPlayerPrivateGStreamer::maxMediaTimeSeekable const): ditto.
(WebCore::MediaPlayerPrivateGStreamer::totalBytes const): ditto.
(WebCore::MediaPlayerPrivateGStreamer::platformDuration const): Return invalid time when the duration is unknown and we&apos;re completely sure that the video isn&apos;t live.
(WebCore::MediaPlayerPrivateGStreamer::handleMessage): Migrate to optional m_isLiveStream preserving the previous behaviour (unset means false).
(WebCore::MediaPlayerPrivateGStreamer::updateStates): ditto.
(WebCore::MediaPlayerPrivateGStreamer::updateDownloadBufferingFlag): ditto.
(WebCore::MediaPlayerPrivateGStreamer::canSaveMediaData const): ditto.
(WebCore::MediaPlayerPrivateGStreamer::triggerRepaint): ditto.
(WebCore::MediaPlayerPrivateGStreamer::movieLoadType const): ditto.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h: m_isLiveStream is now std::optional&lt;bool&gt;.

Canonical link: <a href="https://commits.webkit.org/254682@main">https://commits.webkit.org/254682@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/812f4dbced8470fb87ab8a49557b936bc662d6d0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/89684 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/34235 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/20396 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/98987 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/155803 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/93692 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/32729 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/28202 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/82052 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/93382 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/95331 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/26004 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/76534 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/25965 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/80881 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/80762 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/68956 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/81349 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/30483 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/14842 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/76206 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/30233 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/15776 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/26425 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/3287 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/33681 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/38719 "Found 1 new test failure: imported/w3c/web-platform-tests/html/semantics/embedded-content/the-iframe-element/sandbox-inherit-to-blank-document-unsandboxed.html (failure)") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/78792 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/32392 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/34868 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/17272 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

calvaris commented on Source/WTF/wtf/PrintStream.h:
```diff
@@ -45,6 +45,11 @@ inline const char* boolForPrinting(bool value)
     return value ? "true" : "false";
 }
 
+inline const char* boolForPrinting(const std::optional<bool>&value)
```
I'm sorry to inform you that you have to build 1800 source files again, but you need to add a space between & and value

eocanha replied:
<img src="https://imgs.xkcd.com/comics/compiling.png" />

calvaris commented on Source/WTF/wtf/PrintStream.h:
```diff
@@ -45,6 +45,11 @@ inline const char* boolForPrinting(bool value)
 
+inline const char* boolForPrinting(const std::optional<bool>&value)
+{
+    return value ? (value.value() ? "true" : "false") : "<nullopt>";
```
You can call `boolForPrinting(value.value())` and remove the ()

eocanha committed:
[GStreamer] Fix media player's duration calculation
https://bugs.webkit.org/show_bug.cgi?id=245148

Reviewed by Xabier Rodriguez-Calvar.

Videos without an initially known duration, such as https://traffic.megaphone.fm/GLT1153394917.mp3?updated=1617734730,
are currently detected as infinite videos (live streams). They should be detected as
fixed-length videos instead of as a live broadcast.

This patch changes the behaviour to return invalid time when the video is known not
to be live.

Based on a patch by: Pawel Lampe <pawel.lampe@gmail.com>
See: https://github.com/WebPlatformForEmbedded/WPEWebKit/pull/923

* Source/WTF/wtf/PrintStream.h:
(WTF::boolForPrinting): Added overloaded function to print std::optional<bool> values.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.cpp:
(WebCore::MediaPlayerPrivateGStreamer::seek): Migrate to optional m_isLiveStream preserving the previous behaviour (unset means false).
(WebCore::MediaPlayerPrivateGStreamer::setRate): ditto.
(WebCore::MediaPlayerPrivateGStreamer::setPreload): ditto.
(WebCore::MediaPlayerPrivateGStreamer::buffered const): ditto.
(WebCore::MediaPlayerPrivateGStreamer::maxMediaTimeSeekable const): ditto.
(WebCore::MediaPlayerPrivateGStreamer::totalBytes const): ditto.
(WebCore::MediaPlayerPrivateGStreamer::platformDuration const): Return invalid time when the duration is unknown and we're completely sure that the video isn't live.
(WebCore::MediaPlayerPrivateGStreamer::handleMessage): Migrate to optional m_isLiveStream preserving the previous behaviour (unset means false).
(WebCore::MediaPlayerPrivateGStreamer::updateStates): ditto.
(WebCore::MediaPlayerPrivateGStreamer::updateDownloadBufferingFlag): ditto.
(WebCore::MediaPlayerPrivateGStreamer::canSaveMediaData const): ditto.
(WebCore::MediaPlayerPrivateGStreamer::triggerRepaint): ditto.
(WebCore::MediaPlayerPrivateGStreamer::movieLoadType const): ditto.
* Source/WebCore/platform/graphics/gstreamer/MediaPlayerPrivateGStreamer.h: m_isLiveStream is now std::optional<bool>.

Canonical link: https://commits.webkit.org/254682@main
